src/cns/CnsRegistryReader.ts

```diff
@@ -0,0 +1,201 @@
+import Contract from '../utils/contract';
+
+/** @internal */
+export default class CnsRegistryReader extends EthereumNamingService {
+  readonly registryAddress?: string;
+
+  constructor(source: SourceDefinition = {}) {
+    super(source, NamingServiceName.CNS);
+    this.registryAddress = source.registry;
+    if (this.registryAddress) {
+      this.registryContract = this.buildContract(
+        cnsInterface,
+        this.registryAddress,
+      );
+    }
+  }
+
+  isSupportedDomain(domain: string): boolean {
+    return (
+      domain === 'crypto' ||
+      (domain.indexOf('.') > 0 &&
+        /^.{1,}\.(crypto)$/.test(domain) &&
+        domain.split('.').every((v) => !!v.length))
+    );
+  }
+
+  /**
+   * Resolves the given domain.
+   * @deprecated
+   * @param domain - domain name to be resolved
+   * @returns- Returns a promise that resolves in an object
+   */
+  async resolve(_: string): Promise<ResolutionResponse> {
+    throw new Error('This method is unsupported for CNS');
+  }
+
+  /**
+   * Resolves domain to a specific cryptoAddress
+   * @param domain - domain name to be resolved
+   * @param currencyTicker currency ticker such as
+   *  - ZIL
+   *  - BTC
+   *  - ETH
+   * @returns - A promise that resolves in a string
+   */
+  async address(domain: string, currencyTicker: string): Promise<string> {
+    const tokenId = this.namehash(domain);
+    const ownerPromise = this.owner(domain);
+    const resolver = await this.getResolver(tokenId);
+    if (isNullAddress(resolver)) {
+      await this.throwOwnershipError(domain, ownerPromise);
+    } else {
+      ownerPromise.catch(() => { });
+    }
+    const addr: string | undefined = await this.ignoreResolutionError(
+      ResolutionErrorCode.RecordNotFound,
+      this.fetchAddress(resolver, this.namehash(domain), currencyTicker),
+    );
+    if (!addr) {
+      throw new ResolutionError(ResolutionErrorCode.UnspecifiedCurrency, {
+        domain,
+        currencyTicker,
+      });
+    }
+    return addr;
+  }
+
+  /**
+   * @internal
+   * @param resolver - Resolver address
+   * @param tokenId - namehash of a domain name
+   */
+  private async fetchAddress(
+    resolver: string,
+    tokenId: nodeHash,
+    coinName: string,
+  ): Promise<string> {
+    const resolverContract = this.buildContract(resolverInterface, resolver);
+    const addrKey = `crypto.${coinName.toUpperCase()}.address`;
+    const addr: string = await this.getRecord(resolverContract, 'get', [
+      addrKey,
+      tokenId,
+    ]);
+    return addr;
+  }
+
+  /**
+   * Produces CNS namehash
+   * @param domain - domain to be hashed
+   * @returns CNS namehash of a domain
+   */
+  namehash(domain: string): nodeHash {
+    this.ensureSupportedDomain(domain);
+    return hash(domain);
+  }
+
+  /**
+   * Returns the childhash
+   * @param parent - nodehash of a parent
+   * @param label - child
+   */
+  childhash(_: nodeHash, __: string): nodeHash {
```

Comment:
oh.. how come did we get lower camelcase type names? D:
(rhetorical question)

Reply:
this is special trick ;)